File: android/src/main/java/com/snapchatcamerakit/camera/CameraKitView+Effects.kt
package com.snapchatcamerakit.camera

import com.snapchatcamerakit.camera.errors.CameraNotReadyError
import com.snapchatcamerakit.camera.errors.CameraKitLensNotFoundError
import com.snapchatcamerakit.camera.models.Lens
import com.snapchatcamerakit.utils.DataUtils
import com.snapchatcamerakit.camera.models.CustomLensLaunchData
import com.facebook.react.bridge.*
import com.snap.camerakit.lenses.LensesComponent
import com.snap.camerakit.lenses.whenHasSome

fun CameraKitView.applyLensById(lensId: String, lensGroups: ReadableArray, launchDataMap: ReadableMap?, promise: Promise?) {
    if (this.cameraKitSession == null) {
        throw CameraNotReadyError()
    }

    val groupIds = DataUtils.readableArrayToStringSet(lensGroups)

    this.cameraKitSession!!.lenses.repository.get(LensesComponent.Repository.QueryCriteria.Available(groupIds)) { result ->
        result.whenHasSome { lenses ->
            val lens: LensesComponent.Lens? = lenses.find { _lens -> _lens.id == lensId }
                ?: throw CameraKitLensNotFoundError(id = lensId)

            if (lens != null) {
                val launchData = if (launchDataMap == null) CustomLensLaunchData.empty else CustomLensLaunchData.fromBridge(launchDataMap)

                this.cameraKitSession!!.lenses.processor.apply(lens, launchData) { status ->
                    promise?.resolve(status)
                }
            } else {
                promise?.reject(CameraKitLensNotFoundError(lensId))
            }
        }
    }
}

fun CameraKitView.getLensesByGroupId(lensGroups: ReadableArray, callback: Callback) {
    if (this.cameraKitSession == null) {
        throw CameraNotReadyError()
    }

    val groupIds = DataUtils.readableArrayToStringSet(lensGroups)

    val list = Arguments.createArray()

    this.cameraKitSession!!.lenses.repository.get(LensesComponent.Repository.QueryCriteria.Available(groupIds)) { result ->
        result.whenHasSome { lenses ->
            lenses.forEach() { lens ->
                val lensInstance = Lens(lens)
                list.pushMap(lensInstance.toBridge())
            }
        }

        callback(list, null)
    }
}

fun CameraKitView.clearLenses(callback: Callback) {
    if (this.cameraKitSession == null) {
        throw CameraNotReadyError()
    }

    this.cameraKitSession!!.lenses.processor.clear { success ->
      callback(success, null)
    }
}

File: ios/Camera Kit/CameraKitView+Effects.swift
//
//  CameraKitView+Effects.swift
//  react-native-snapchat-camera-kit
//
//  Created by Rıdvan Altun on 28.03.2023.
//

import SCSDKCameraKit

extension CameraKitView {
  func applyLensById(lensId: NSString, lensGroups: NSArray, launchDataMap: NSDictionary?, promise: Promise?) {
    var lens: Lens?

    lensGroups.forEach { groupID in
      // swiftlint:disable force_cast
      lens = self.cameraController?.cameraKit.lenses.repository.lens(id: lensId as String, groupID: groupID as! String)
      // swiftlint:enable force_cast
    }

    if lens != nil {
      var launcData: LensLaunchData?

      if launchDataMap != nil {
        launcData = CustomLensLaunchData.fromBridge(data: launchDataMap!)
      }

      cameraController?.applyLens(lens!, launchData: launcData) { status in
        promise?.resolve(status)
      }
    } else {
      // TODO: lens not found
    }
  }

  func getLensesByGroupId(lensGroups: NSArray, callback jsCallbackFunc: @escaping RCTResponseSenderBlock) {
    let lenses: NSMutableArray = []
    let err = NSNull()

    lensGroups.forEach { groupId in
      // swiftlint:disable force_cast
      self.cameraController?.cameraKit.lenses.repository.lenses(groupID: groupId as! String).forEach { lens in
        let lensObject = LensModel(lens: lens).toBridge()

        lenses.add(lensObject)
      }
      // swiftlint:enable force_cast
    }

    jsCallbackFunc([lenses as? [Any]? as Any, err])
  }

  func clearLenses(callback jsCallbackFunc: @escaping RCTResponseSenderBlock) {
    let err = NSNull()

    cameraController?.clearLens { status in
      jsCallbackFunc([status, err])
    }
  }
}
